File: internal/router/embedding.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2025-2026 lin-snow

package router

import (
	"net/http"

	"github.com/danielgtaylor/huma/v2"
	"github.com/lin-snow/ech0/internal/handler"
	authModel "github.com/lin-snow/ech0/internal/model/auth"
	authService "github.com/lin-snow/ech0/internal/service/auth"
)

// registerEmbeddingHuma 注册 Embedding 向量索引操作（owner / 管理员，需 admin:settings scope）。
// 注意：Embedding 设置（get/update）仍归口到 registerSettingHuma（尚未迁移）。
func registerEmbeddingHuma(api huma.API, h *handler.Bundle, revoker authService.TokenRevoker) {
	register(api, secured(revoker, authModel.ScopeAdminSettings), huma.Operation{
		OperationID: "embedding-reindex",
		Method:      http.MethodPost,
		Path:        "/embedding/reindex",
		Summary:     "触发全量向量索引重建",
		Description: "提交一次全量向量索引回填作业，起即返回（异步）。",
		Tags:        []string{"Embedding"},
	}, h.EmbeddingHandler.Reindex)

	register(api, secured(revoker, authModel.ScopeAdminSettings), huma.Operation{
		OperationID: "embedding-reindex-status",
		Method:      http.MethodGet,
		Path:        "/embedding/reindex/status",
		Summary:     "查询重建索引作业状态",
		Description: "前端按类型轮询；查无作业行时返回 status=idle。",
		Tags:        []string{"Embedding"},
	}, h.EmbeddingHandler.ReindexStatus)

	register(api, secured(revoker, authModel.ScopeAdminSettings), huma.Operation{
		OperationID: "embedding-reindex-cancel",
		Method:      http.MethodPost,
		Path:        "/embedding/reindex/cancel",
		Summary:     "取消进行中的重建索引作业",
		Description: "取消后返回最新状态（轮询收敛到 cancelled）。",
		Tags:        []string{"Embedding"},
	}, h.EmbeddingHandler.CancelReindex)
}

File: web/src/service/api/embedding.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2025-2026 lin-snow

import { request } from '../request'

// 获取 Embedding 向量设置
export function fetchGetEmbeddingSettings() {
  return request<App.Api.Embedding.EmbeddingSetting>({
    url: '/embedding/settings',
    method: 'GET',
  })
}

// 更新 Embedding 向量设置
export function fetchUpdateEmbeddingSettings(data: App.Api.Embedding.EmbeddingSettingDto) {
  return request<null>({
    url: '/embedding/settings',
    method: 'PUT',
    data,
  })
}

// 提交全量重建向量索引作业（异步：起即返回作业状态，前端轮询进度）
export function fetchReindexEmbeddings() {
  return request<App.Api.Embedding.ReindexStatus>({
    url: '/embedding/reindex',
    method: 'POST',
  })
}

// 查询重建索引作业状态（按 type 轮询，无需 id）
export function fetchReindexStatus() {
  return request<App.Api.Embedding.ReindexStatus>({
    url: '/embedding/reindex/status',
    method: 'GET',
  })
}

// 取消进行中的重建索引作业
export function fetchCancelReindex() {
  return request<App.Api.Embedding.ReindexStatus>({
    url: '/embedding/reindex/cancel',
    method: 'POST',
  })
}
